module/os/linux/zfs/vdev_disk.c

```diff
@@ -670,8 +590,539 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+{
+	/* Max segments we need in each BIO to take all pages */
+	uint_t max_segs = MIN(pages, vdev_bio_max_segs(bdev));
+
+	/* Number of BIOs we need, at max_segs segments each */
+	int max_bios = (pages + max_segs - 1) / max_segs;
+
+	vbio_t *vbio = kmem_zalloc(sizeof (vbio_t) +
+	    sizeof (struct bio *) * max_bios, KM_SLEEP);
+
+	vbio->vbio_zio = zio;
+	vbio->vbio_bdev = bdev;
+	atomic_set(&vbio->vbio_ref, 0);
+	vbio->vbio_max_segs = max_segs;
+	vbio->vbio_max_bios = max_bios;
+	vbio->vbio_npages = pages;
+	vbio->vbio_cur = 0;
+	vbio->vbio_offset = zio->io_offset;
+
+	return (vbio);
+}
+
+static int
+vbio_add_page(vbio_t *vbio, struct page *page, uint_t size, uint_t offset)
+{
+	struct bio *bio;
+
+	/*
+	 * Weird housekeeping error; shouldn't happen, but try and save the
+	 * furniture if it happens in production.
+	 */
+	ASSERT3U(vbio->vbio_npages, >, 0);
+	if (vbio->vbio_npages == 0)
+		return (SET_ERROR(ENOMEM));
+
+	for (;;) {
+		bio = vbio->vbio_bio[vbio->vbio_cur];
+		if (bio == NULL) {
+			/* New BIO, allocate and set up */
+			bio = vdev_bio_alloc(vbio->vbio_bdev, GFP_NOIO,
+			    MIN(vbio->vbio_npages, vbio->vbio_max_segs));
+			if (unlikely(bio == NULL))
+				return (SET_ERROR(ENOMEM));
+			BIO_BI_SECTOR(bio) = vbio->vbio_offset >> 9;
+			vbio->vbio_bio[vbio->vbio_cur] = bio;
+			vbio->vbio_nbios++;
+		}
+
+		if (bio_add_page(bio, page, size, offset) == size) {
+			vbio->vbio_npages--;
+			return (0);
+		}
+
+		/* No room, set up for a new BIO and loop */
+		vbio->vbio_offset += BIO_BI_SIZE(bio);
+		vbio->vbio_cur++;
+
+		VERIFY3U(vbio->vbio_cur, <, vbio->vbio_max_bios);
+	}
+}
+
+BIO_END_IO_PROTO(vdev_disk_io_rw_completion, bio, error);
+static void vbio_put(vbio_t *vbio);
+
+static void
+vbio_submit(vbio_t *vbio, int flags)
+{
+	struct blk_plug plug;
+
+	/*
+	 * Take a reference for each BIO we're about to submit, plus one to
+	 * protect us from BIOs completing before we're done submitting them
+	 * all, causing vbio_put() to free vbio out from under us and/or the
+	 * zio to be returned before all its IO has completed.
+	 */
+	atomic_set(&vbio->vbio_ref, vbio->vbio_nbios + 1);
+
+	/*
+	 * If we're submitting more than one BIO, inform the block layer so
+	 * it can batch them if it wants.
+	 */
+	if (vbio->vbio_nbios > 1)
+		blk_start_plug(&plug);
+
+	/* Submit all the BIOs */
+	for (int i = 0; i < vbio->vbio_nbios; i++) {
+		struct bio *bio = vbio->vbio_bio[i];
+		ASSERT3P(bio, !=, NULL);
+
+		bio->bi_end_io = vdev_disk_io_rw_completion;
+		bio->bi_private = vbio;
+		bio_set_op_attrs(bio,
+		    vbio->vbio_zio->io_type == ZIO_TYPE_WRITE ?
+		    WRITE : READ, flags);
+		vdev_submit_bio(bio);
+	}
+
+	/* Finish the batch */
+	if (vbio->vbio_nbios > 1)
+		blk_finish_plug(&plug);
+
+	/* Release the extra reference */
+	vbio_put(vbio);
+}
+
+static void
+vbio_return_abd(vbio_t *vbio)
+{
+	zio_t *zio = vbio->vbio_zio;
+	if (vbio->vbio_abd == NULL || vbio->vbio_abd == zio->io_abd)
```

Comment:
Isn't the second condition excessive after being checked in vdev_disk_io_rw() on line 988?

Reply:
Yeah, good spot. It was left over from a previous iteration. Removed.